scripts/table_updates/utilities.py

```diff
@@ -126,12 +186,95 @@ def revert_table_version(syn, table_id):
+    table_columns = list(set(table_columns) & set(old_data.columns))
     temp_data = old_data[table_columns]
     table_query = syn.tableQuery("SELECT * from %s" % table_id)
     syn.delete(table_query.asRowSet())
-    syn.store(Table(table_schema, temp_data))
\ No newline at end of file
+    syn.store(Table(table_schema, temp_data))
+
+
+def update_tier1a(syn: synapseclient.Synapse, form: str, master_table: pandas.DataFrame, main_genie_table: pandas.DataFrame, column_mapping_table: pandas.DataFrame, bpc_column_list: List[str],logger: logging.Logger = None, cohort: str = "") -> Tuple[str, pandas.DataFrame]:
+    """Replace tier1a variables in patient_characteristics or cancer_panel_test table with Main GENIE release files
+
+    Args:
+        syn (synapseclient.Synapse): The synapse client connection
+        form (str): The form name, can be either patient_characteristics or cancer_panel_test
+        master_table (pandas.DataFrame): Table of all of the primary or irr BPC tables
+        main_genie_table (pandas.DataFrame): The dataframe of Main GENIE release
+        column_mapping_table (pandas.DataFrame): The column mapping table between BPC and Main GENIE
+        bpc_column_list (List[str]): The column list to be replaced
+        logger (logging.Logger, optional): The custom logger. Optional. Defaults to None.
+        cohort (str, optional): The cohort name. Defaults to "".
+
+    Returns:
+        Tuple[str, pandas.DataFrame]: The synapse ID for BPC table to be modified and the updated table as dataframe
+    """
+    # check the validity of bpc_column_list
+    valid_col = column_mapping_table.loc[column_mapping_table["prissmm_form"] == form,].prissmm_element.tolist()
+    assert all(item in valid_col for item in bpc_column_list), (f"Invalid bpc_column_list. Column names should be matching {valid_col}.")
+
+    logger.info(f"Update {bpc_column_list} in {form}")
+    # load bpc table
+    cpt_table_id = master_table.loc[
+        master_table["form"] == form, "id"
+    ].values[0]
+
+    if cohort:
+        condition = f"cohort = '{cohort}'"
+        cpt_dat = download_synapse_table(syn, cpt_table_id, condition = condition)
+    else:
+        cpt_dat = download_synapse_table(syn, cpt_table_id)
+    cpt_dat.index = cpt_dat.index.map(str)
+    cpt_dat["index"] = cpt_dat.index
+    # subset main_genie_table based on bpc_column_list
+    main_genie_column_list = [", ".join(column_mapping_table.loc[column_mapping_table["prissmm_element"] == col,].genie_element) for col in bpc_column_list]
+    
+    if form == "patient_characteristics":
+        main_genie_table = main_genie_table[main_genie_column_list + ["PATIENT_ID"]]
+        cpt_seq_dat = cpt_dat.merge(
+            main_genie_table,
+            how="left",
+            left_on="genie_patient_id",
+            right_on="PATIENT_ID",
+        )
+    else: 
+        main_genie_table = main_genie_table[main_genie_column_list + ["SAMPLE_ID"]]
```

Comment:
I'm not sure what the original code looked like but was there ever a handling of potential duplicates after merging?

Reply:
The original code doesn't handle duplicate.